scipy/linalg/_solve_toeplitz.pyx

```diff
@@ -0,0 +1,129 @@
+# Author: Robert T. McGibbon, December 2014
+#
+# cython: boundscheck=False, wraparound=False, cdivision=True
+from numpy import zeros, asarray, complex128, float64
+from numpy.linalg import LinAlgError
+from numpy cimport npy_intp, complex128_t, float64_t
+
+
+cdef fused dz:
+    float64_t
+    complex128_t
+
+
+def levinson(dz[::1] a, dz[::1] b):
+    """Solve a linear Toeplitz system using Levinson recursion.
+
+    Parameters
+    ----------
+    a : array, dtype=double or complex128, shape=(2n-1,)
+        The first column of the matrix in reverse order (without the diagonal)
+        followed by the first (see below)
+    b : array, dtype=double  or complex128, shape=(n,)
+        The right hand side vector. Both a and b must have the same type
+        (double or complex128).
+
+    Notes
+    -----
+    For example, the 5x5 toeplitz matrix below should be represented as
+    the linear array ``a`` on the right ::
+
+        [ a0    a1   a2  a3  a4 ]
+        [ a-1   a0   a1  a2  a3 ]
+        [ a-2  a-1   a0  a1  a2 ] -> [a-4  a-3  a-2  a-1  a0  a1  a2  a3  a4]
+        [ a-3  a-2  a-1  a0  a1 ]
+        [ a-4  a-3  a-2  a-1 a0 ]
+
+    Returns
+    -------
+    x : arrray, shape=(n,)
+        The solution vector
+    reflection_coeff : array, shape=(n+1,)
+        Toeplitz reflection coefficients. When a is symmetric Toeplitz and
+        ``b`` is ``a[n:]``, as in the solution of autoregressive systems,
+        then ``reflection_coeff`` also correspond to the partial
+        autocorrelation function.
+    """
+    # Adapted from toeplitz.f90 by Alan Miller, accessed at
+    # http://jblevins.org/mirror/amiller/toeplitz.f90
+    # Released under a Public domain declaration.
+
+    if a.itemsize == 8:
+        # Is this the best way to check the dtype of the memoryview?
+        assert b.itemsize == 8
```

Comment:
Yeah, I didn't have a lot of experience with fused types. Your recommended solution using `is` is much cleaner.